staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package shufflesharding
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"strings"
+)
+
+const maxHashBits = 60
+
+// ValidateParameters finds errors in the parameters for shuffle
+// sharding.  Returns a slice for which `len()` is 0 if and only if
+// there are no errors.  The entropy requirement is evaluated in a
+// fast but approximate way: bits(deckSize^handSize).
+func ValidateParameters(deckSize, handSize int) (errs []string) {
```

Comment:
This function was intended to cover all the restrictions in https://github.com/kubernetes/kubernetes/blob/feature-rate-limiting/pkg/apis/flowcontrol/validation/validation.go#L221-L233 .  But it is actually only used for one test in there, because the list returned from this function is not adequately expressive to drive all those validation tests.  It is awkward for https://github.com/kubernetes/kubernetes/blob/feature-rate-limiting/pkg/apis/flowcontrol/validation/validation.go#L231 to invoke this whole function just to get one test.  I suggest that we either (a) change the returned value so that this function can do all of the testing in https://github.com/kubernetes/kubernetes/blob/feature-rate-limiting/pkg/apis/flowcontrol/validation/validation.go#L221-L233 or (b) factor out of this function the one test that the validation code needs.

Reply:
(b) sounds probably better. we factor a `RequiredEntropyBits` out of this and make the new function returning one error.

Reply:
will factor this out, let do the most validations in flowcontrol.